### PR TITLE
KAFKA-6430: Add buffer between Java data stream and gzip stream

### PR DESCRIPTION
**Pull request by ying-zheng**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java [hunk 1]
```diff
@@ -59,7 +61,7 @@ public OutputStream wrapForOutput(ByteBufferOutputStream buffer, byte messageVer
         @Override
         public InputStream wrapForInput(ByteBuffer buffer, byte messageVersion, BufferSupplier decompressionBufferSupplier) {
             try {
-                return new GZIPInputStream(new ByteBufferInputStream(buffer));
+                return new BufferedInputStream(new GZIPInputStream(new ByteBufferInputStream(buffer)), 1 << 14);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: How was the default buffer size chosen?

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. Just one question.

**Review by hachikuji — APPROVED** [review]
LGTM

**Comment by ijuma** [comment]
I'm going to merge this to trunk and 1.1 branches. @ying-zheng, if you have additional feedback, we can address it in a follow-up.

**Review by ijuma — APPROVED** [review]
LGTM, pushed an additional commit with minor changes (already reviewed by @hachikuji).

**Comment by ying-zheng** [comment]
@ijuma thank you for reviewing and fixing the PR!

**Pull request merged**